src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,666 @@
+                }
+                code => {
+                    return Err(anyhow::anyhow!(
+                        "server returned unexpected HTTP status code {} for {}\nbody: {}",
+                        code,
+                        self.full_url(path),
+                        str::from_utf8(&result.data).unwrap_or("<invalid utf8>"),
+                    ))
+                    .into();
+                }
+            }
+
+            return Poll::Ready(Ok(LoadResponse::Data {
+                raw_data: result.data,
+                index_version: Some(result.index_version),
+            }));
+        }
+
+        if self.config.offline() {
+            return Poll::Ready(Err(anyhow::anyhow!(
+                "can't download index file from '{}': you are in offline mode (--offline)",
+                self.url
+            )));
+        }
+
+        // Looks like we're going to have to do a network request.
+        self.start_fetch()?;
+
+        // Load the registry config.
+        if self.registry_config.is_none() && path != Path::new("config.json") {
+            match self.config()? {
+                Poll::Ready(_) => {}
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+
+        let mut handle = ops::http_handle(self.config)?;
+        let full_url = self.full_url(path);
+        debug!("fetch {}", full_url);
+        handle.get(true)?;
+        handle.url(&full_url)?;
+        handle.follow_location(true)?;
+
+        // Enable HTTP/2 if possible.
+        if self.multiplexing {
+            handle.http_version(HttpVersion::V2)?;
+        } else {
+            handle.http_version(HttpVersion::V11)?;
+        }
+
+        // This is an option to `libcurl` which indicates that if there's a
+        // bunch of parallel requests to the same host they all wait until the
+        // pipelining status of the host is known. This means that we won't
+        // initiate dozens of connections to crates.io, but rather only one.
+        // Once the main one is opened we realized that pipelining is possible
+        // and multiplexing is possible with static.crates.io. All in all this
+        // reduces the number of connections done to a more manageable state.
+        handle.pipewait(true)?;
+
+        // Make sure we don't send data back if it's the same as we have in the index.
+        let mut headers = List::new();
+        if let Some(index_version) = index_version {
+            if let Some((key, value)) = index_version.split_once(':') {
+                match key {
+                    ETAG => headers.append(&format!("If-None-Match: {}", value.trim()))?,
+                    LAST_MODIFIED => {
+                        headers.append(&format!("If-Modified-Since: {}", value.trim()))?
+                    }
+                    _ => debug!("unexpected index version: {}", index_version),
+                }
+            }
+        }
+        handle.http_headers(headers)?;
```

Comment:
So if I'm understanding correctly, `index_version` for a registry is registry defined?

This was something that wasn't obvious to me and took some digging.  I had assumed this was a schema version number rather than a data revision number.

Digging in some more, it seems like this is pre-established within cargo.

Reply:
`index_version` is used for cache validation.

For git-based registries `index_version` is the commit id of the HEAD of the repo when the cache file was created. This allows cargo to skip talking to libgit2 when the git repo hasn't changed.

For http-based registries `index_version` is a string `ETag: <etag value>` or `Last-Modified: <last-modified value>` or `Unknown` (in that preference order) based on what the server sent when fetching the index. This allows cargo to send `If-None-Match` or `If-Modified-Since` headers and receive an HTTP 304 (not modified) when fetching metadata. 

The cache in http mode also allows `--offline` mode to work (where the cache is used regardless of the `index_version`)

Reply:
The naming may no longer match how we use it. If you can think of a better name now would be a great time to change it.

Reply:
> The naming may no longer match how we use it

The schema versioning added in https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html#index-changes is what led to my confusion.  it also doesn't help that we are storing structured data in a string.

As for names, its hard to say.
Unfortunately, I can't come up with a clear winner.  The best I can come up with is `index_digest` as that doesn't convey an `Ord` version scheme but a `Eq` identifier.

> If you can think of a better name now would be a great time to change it.

For me, I disagree as making unrelated improvements to a PR can derail it and make it harder to review.  Soon after this PR, yes.

Reply:
Do people like any of these names?
`cache_key`
`cache_identifier`
`cache_identity`
`cache_version`
`cache_tag`
`index_cache_<one of the above options>`